Add repo state and possibility to specify package name

javierbertoli opened the pull request:
This PR tried to address #150, adding a separate state to manage repos `repo.sls`.

When testing, found many issues due to old/deprecated versions/repos, so I tried to refactor the formula to be more up-to-date:

* Added support for testing on:
  * ubuntu-1804
  * debian-stretch
  * debian-jessie

* Added `osfamilymap.yaml`
* Updated repo info

This probably can fix/close #109, #150, #131, #136 and #95 

I'm not using this formula heavily, so if you can check this PR and make suggestions, any comment will be welcome.

It seems that a lot of refactoring is needed to clean old stuff and perhaps this PR helps to start doing it :smile:

noelmcloughlin commented on test/integration/version-18.03.1/testinfra/test_docker.py:
```diff
@@ -2,9 +2,9 @@
 
 def test_package_is_installed(Package):
-    docker = Package('docker-engine')
+    docker = Package('docker-ce')
```
Could the package name be configurable?

noelmcloughlin replied:
See #153 could you use `docker:pkg:name` here.

blbradley replied:
Can't use a saltstack pillar var in test python code without non-trivial work

noelmcloughlin commented on test/integration/version-18.03.1/testinfra/test_docker.py:
```diff
@@ -2,9 +2,9 @@
+    docker = Package('docker-ce')
     assert docker.is_installed
-    assert docker.version.startswith('1.12.2')
+    assert docker.version.startswith('18.03.1')
```
Could the package version be configurable?

noelmcloughlin replied:
See #153 could you use `docker:pkg:version` here (and elsewhere) please?

blbradley replied:
Can't use a saltstack pillar var in test python code without non-trivial work

noelmcloughlin replied:
Hi @javierbertoli @blbradley I'm okay with not complicating things so leave this as is!! thx

blbradley commented on .travis.yml:
```diff
@@ -9,4 +9,17 @@ services:
 
-script: bundle exec kitchen verify
+
+env:
```
Are you sure you need this? I think the matrix is controlled by suiteXplatform in Test Kitchen.

blbradley commented:
I don't really use this formula anymore but will give some suggestions.

blbradley commented:
What version of Docker does the default suite test these days?

blbradley commented on .travis.yml:
```diff
@@ -9,4 +9,17 @@ services:
+    - INSTANCE: default-debian-stretch
+    - INSTANCE: default-debian-jessie
+    - INSTANCE: default-centos-7
+    - INSTANCE: version-1122-ubuntu-1804
```
Hey I think these names need to start version-18031 now. Right? If you all choose to keep this. This is one reason I let Test Kitchen just control the test matrix.

blbradley replied:
Yep! See the errors in Travis:

`No instances for regex `version-1122-ubuntu-1804', try running `kitchen list'`

blbradley commented:
Thanks for the quality updates, @javierbertoli. I won't give a 👍  or 👎 on this because I'm not a user of this formula anymore. But I'm happy to answer any questions about the test suite if you hit any trouble.

noelmcloughlin commented:
Hi @javierbertoli Do you need me to work with you on this?

javierbertoli commented:
Hey, sorry. I'm having some unrelated, Real Life (TM) issues that left me litte time to advance on this :yum: 

I have some commits that failed local tests, that I need to review and push. I'll see if I can push them today/tomorrow. Is it ok? Or do you prefer I push them as they are, so we can review them?

noelmcloughlin commented:
@javierbertoli  No worries, in your own time!!! Good luck

javierbertoli commented:
@noelmcloughlin, @blbradley,  updated the PR. Had to remove testing on CentOS, as docker-in-docker fails. Updated to test Debian 8/9 & Ubuntu 18.04.

To keep things backward-ish compatible, kept the `use_old_repo` parameter & repo, but kept it somewhat 'hardcoded', as the new format is capable of managing all the parameters (repo & package) to install any package from any repo. I think in a future iteration we can remove the old parameters.

I think the formula needs a some cleanup and refactoring (ie, I think that `kernels.sls` can probably be removed now), but that can be done in future PRs.

Please let me know if you have any suggestions or what anything else done in the PR.

noelmcloughlin commented on docker/kernel.sls:
```diff
@@ -1,10 +1,11 @@
+{%- if docker.kernel is defined and grains['os_family']|lower == 'debian' %}
 pkgrepo dependencies:
   pkg.installed:
     - name: python-apt
```
Good fix. There is no `python-apt` package on non-Debian distros so state fails.

noelmcloughlin commented:
Should we just dump the `use_old_repo` support in this PR - because nobody needs this?

javierbertoli commented:
Don't know. If you ask me, I'd dump it, same with the `kernel.sls` state file, as those things are not needed anymore. But didn't want to add too much refactoring in this PR (as it already is, is quite big), so I'm OK with adding a few follow ups to clean the formula. Wdyt?

noelmcloughlin commented:
Yes - followup would be appreciated if someone can make time.  This LGTM as is, and future PR can progress things. Thanks @javierbertoli

aboe76 commented:
@javierbertoli and @noelmcloughlin merged it.
